### PR TITLE
node: try using static Vector limits

### PR DESCRIPTION
**Pull request by slack-coder**
Support clearly defining message constraints by introducing a new vector type.  The vector type takes a statically defined upper bound as a Type parameter.

Should we pursue using this?  The code is to express the idea and not meant as a final version, some clean up will be in order.

```
pub struct InventoryAnnouncement {
    /// Node inventory.
    pub inventory: BoundedVec<INVENTORY_LIMIT, Id>,
    /// Time of announcement.
    pub timestamp: Timestamp,
}
```

### CODE REVIEW

**Comment by FintanH** [comment]
I'm into representing these things at the type level :grin: 

It was also used in `radicle-link` and there was a helper crate for these things https://github.com/radicle-dev/radicle-link/blob/master/data/src/bounded.rs. You might want to use that.

**Comment by dr-orlovsky** [comment]
I think I have guys what you need here: https://github.com/rust-amplify/rust-amplify/pull/155

@cloudhead was going to use that amplify crate in anyway as a dependency.

**Comment by cloudhead** [comment]
Yeah at the very list we should be using `const` generics for the size limits, not regular generics. The `Confined` type seems pretty much what we need, if we're going to be depending on `amplify` anyway after the transport merge..

**Comment by dr-orlovsky** [comment]
@cloudhead I have completed and released that implementation as a part of `amplify = 4.0.0-alpha.1`. You can see the whole final implementation here: https://github.com/rust-amplify/rust-amplify/blob/master/src/confinement.rs

The list of exported simplified newtype names to drop-in usage is this:
https://docs.rs/amplify/4.0.0-alpha.1/amplify/confinement/index.html#types

https://github.com/rust-amplify/rust-amplify/blob/master/src/confinement.rs#L677-L752

**Comment by slack-coder** [comment]
I've revamped the custom type with some inspiration from radicle-links type.  I opted for the custom type because importing the module to use const generics was more difficult than expected.

A separate PR can be set up for changing the bounded vector type if and when necessary.

File: radicle-node/src/service.rs
Comment (cloudhead) [inline radicle-node/src/service.rs:1330]:
```suggestion
                "inventory announcement limit ({}) exceeded, other nodes will see only some of your projects",
```

File: radicle-node/src/service/message.rs
Comment (cloudhead) [inline radicle-node/src/service/message.rs:528]: Can you add an assert for decoding too?

File: radicle-node/src/tests.rs
Comment (cloudhead) [inline radicle-node/src/tests.rs:618]:
```suggestion
```

File: radicle-node/src/wire/message.rs
Comment (cloudhead) [inline radicle-node/src/wire/message.rs:21]: :eyes:

File: radicle-node/src/service.rs
Comment (cloudhead) [inline radicle-node/src/service.rs:1307]:
```suggestion
            .expect("external addresses are within the limit");
```

File: radicle-node/src/service.rs
Comment (cloudhead) [inline radicle-node/src/service.rs:1288]:
```suggestion
            .expect("external addresses are within the limit");
```

File: radicle-node/src/service/message.rs
Comment (slack-coder) [reply radicle-node/src/service/message.rs:528]: Sure, though does the property test for encoding/decoding cover this already?

https://github.com/radicle-dev/heartwood/blob/master/radicle-node/src/wire/message.rs#L405

Comment (slack-coder) [reply radicle-node/src/service/message.rs:528]: In fact, the property test should cover both cases already right?

File: radicle-node/src/tests.rs
Comment (slack-coder) [reply radicle-node/src/tests.rs:618]: Looks like the constant was not needed.

**Comment by slack-coder** [comment]
The code has been updated as suggested

File: radicle-node/src/service/message.rs
Comment (cloudhead) [reply radicle-node/src/service/message.rs:528]: It doesn't (see unwrap) -- it only produces valid messages.

Comment (slack-coder) [reply radicle-node/src/service/message.rs:528]: I went ahead and implemented it while waiting on the answer.

Comment (cloudhead) [reply radicle-node/src/service/message.rs:528]: I must be missing something, I don't see where this is tested?

Comment (slack-coder) [reply radicle-node/src/service/message.rs:528]: Indeed, its annoying.  Clicking on the file link of the comment points to a specific commit.   I thought I forgot to push the changes.  Maybe because the comment was on the specific commit?

Here are the relevant changes: https://github.com/radicle-dev/heartwood/pull/115/commits/0f7543799bd9c333344e20091a96bb3456638f3b

Comment (cloudhead) [reply radicle-node/src/service/message.rs:528]: I saw those changes, but they don't test that decoding fails with a too big payload.

Comment (cloudhead) [reply radicle-node/src/service/message.rs:528]:
```rust
        assert!(
            msg.encode(&mut buf).is_ok(),
            "INVENTORY_LIMIT is too big to support message encoding",
        );
```
The message here also doesn't make sense, it should say `"the inventory size is smaller than the limit"`

**Review by cloudhead — APPROVED** [review]
Perfect.

**Pull request merged**
